spring-cloud-stream-integration-tests/src/test/java/org/springframework/cloud/stream/config/StreamListenerMethodTests.java

```diff
@@ -0,0 +1,399 @@
+/*
+ * Copyright 2016 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.stream.config;
+
+import java.util.UUID;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+
+import org.junit.Test;
+
+import org.springframework.beans.factory.BeanCreationException;
+import org.springframework.boot.SpringApplication;
+import org.springframework.boot.autoconfigure.EnableAutoConfiguration;
+import org.springframework.cloud.stream.annotation.EnableBinding;
+import org.springframework.cloud.stream.annotation.Input;
+import org.springframework.cloud.stream.annotation.Output;
+import org.springframework.cloud.stream.annotation.StreamListener;
+import org.springframework.cloud.stream.messaging.Processor;
+import org.springframework.cloud.stream.messaging.Sink;
+import org.springframework.context.ConfigurableApplicationContext;
+import org.springframework.integration.support.MessageBuilder;
+import org.springframework.messaging.Message;
+import org.springframework.messaging.MessageChannel;
+import org.springframework.messaging.MessageHandler;
+import org.springframework.messaging.MessagingException;
+import org.springframework.messaging.SubscribableChannel;
+import org.springframework.messaging.handler.annotation.Payload;
+import org.springframework.util.Assert;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.junit.Assert.fail;
+
+/**
+ * @author Marius Bogoevici
+ * @author Ilayaperumal Gopinathan
+ */
+public class StreamListenerMethodTests {
+
+	@Test
+	public void testMethodWIthInputOnStreamListener() throws Exception {
+		try {
+			SpringApplication.run(TestMethodWIthInputOnStreamListener.class, "--server.port=0");
```

Comment:
It shouldn't be so complex. All we eventually wanted was to be able to support this (with or without return argument) in addition to what we already supported: 

```
@StreamListener(Processor.INPUT)
@SendTo(Processor.OUTPUT)
public Flux<String> toUpper(Flux<String> input)
```

See my last comment : https://github.com/spring-cloud/spring-cloud-stream/pull/681/files/c05266a3f2471da32159f77553e9cdc1b84f5581#r84973574

Thinking again through all this, the initial test was:

- does StreamListener have an argument ?
a) if not, handle it as a declarative method
b) if yes, handle it as a message handler method;

This becomes:
- does StreamListener have an argument ? 
a) if not, handle it as a declarative method;
b) if yes:
     - if the method doesn't have a single parameter, fail automatically (this should currently happen anyway)
     b1) if the argument is bindable or convertible to bindable, handle it as a declarative method;
     b2) if not, handle it as a message handler method;

We already have a) and b2) implemented (including validations) so all we need is to add an additional test in the branch that tests StreamListener where we validate that we're not in b1 and handle accordingly. So these would be the three variants that work:

```
@StreamListener(Processor.INPUT)
@SendTo(Processor.OUTPUT)
public String> toUpper(String input)
```

```
@StreamListener(Processor.INPUT)
@SendTo(Processor.OUTPUT)
public Flux<String> toUpper(Flux<String> input)
```

```
@StreamListener
@Output(Processor.OUTPUT)
public Flux<String> toUpper(@Input(Processor.Input) Flux<String> input)
```

Where the latter is a particular case of the already supported idiom where you can have an arbitrary number of params annotated with Input and Output and maybe a return value annotated with Output. Would this be too confusing? I think that there is value in aligning the two variants just as the reactive web model allows for @GetMapping and so on.